File: src/main/java/com/herbertgao/telegram/bot/gaokao/business/common/util/GaokaoBotUtil.java
package com.herbertgao.telegram.bot.gaokao.business.common.util;

import com.herbertgao.telegram.bot.gaokao.business.common.constant.TemplateReplace;
import com.herbertgao.telegram.bot.gaokao.database.domain.ExamDate;
import org.apache.commons.lang3.StringUtils;

import java.time.Duration;
import java.time.LocalDateTime;

/**
 * 高考机器人Util
 *
 * @author HerbertGao
 * @date 2019-06-09
 */
public class GaokaoBotUtil {

    /**
     * 是考试开始时间
     *
     * @param exam 考试
     * @param now  现在
     * @return {@link Boolean}
     */
    public static Boolean isExamBeginTime(ExamDate exam, LocalDateTime now) {
        LocalDateTime examBeginDate = exam.getExamBeginDate();
        return examBeginDate.isEqual(now);
    }

    /**
     * 是考试时间
     *
     * @param exam 考试
     * @param now  现在
     * @return {@link Boolean}
     */
    public static Boolean isExamTime(ExamDate exam, LocalDateTime now) {
        LocalDateTime examBeginDate = exam.getExamBeginDate();
        LocalDateTime examEndDate = exam.getExamEndDate();
        return (examBeginDate.isBefore(now) || examBeginDate.isEqual(now)) && (examEndDate.isAfter(now) || examEndDate.isEqual(now));
    }

    /**
     * 是过期的考试
     *
     * @param exam 考试
     * @param now  现在
     * @return {@link Boolean}
     */
    public static Boolean isExpiredExam(ExamDate exam, LocalDateTime now) {
        LocalDateTime examEndDate = exam.getExamEndDate();
        return examEndDate.isBefore(now);
    }

    /**
     * 得到倒计时字符串
     *
     * @param exam     考试
     * @param now      现在
     * @param template 模板
     * @return {@link String}
     */
    public static String getCountDownString(ExamDate exam, LocalDateTime now, String template) {
        if (isExamBeginTime(exam, now)) {
            return exam.getExamDesc() + "开始了！" + System.getProperty("line.separator");
        } if (isExamTime(exam, now)) {
            return exam.getExamDesc() + "正在进行中！" + System.getProperty("line.separator");
        } else if (isExpiredExam(exam, now)) {
            return exam.getExamDesc() + "已经结束了。" + System.getProperty("line.separator");
        } else {
            String rtn = template;
            String[] searchList = {
                    TemplateReplace.EXAM_YEAR,
                    TemplateReplace.EXAM,
                    TemplateReplace.EXAM_S,
                    TemplateReplace.TIME
            };
            String[] replacementList = {
                    exam.getExamYear().toString(),
                    exam.getExamDesc(),
                    exam.getShortDesc(),
                    getCountDownTime(exam, now)
            };
            rtn = StringUtils.replaceEach(rtn, searchList, replacementList);
            rtn += System.getProperty("line.separator");
            return rtn;
        }
    }

    /**
     * 获取倒计时时间文字
     *
     * @param exam
     * @param now
     * @return
     */
    public static String getCountDownTime(ExamDate exam, LocalDateTime now) {
        LocalDateTime examBeginDate = exam.getExamBeginDate();
        Duration between = Duration.between(now, examBeginDate);
        return DateTimeUtils.formatDuration(between);
    }

}

File: internal/util/gaokao_bot_util.go
package util

import (
	"fmt"
	"strings"
	"time"

	"github.com/herbertgao/gaokao_bot/internal/model"
)

// IsExamBeginTime 判断是否考试开始时刻
func IsExamBeginTime(exam *model.ExamDate, now time.Time) bool {
	return now.After(exam.ExamBeginDate) && now.Before(exam.ExamBeginDate.Add(time.Minute))
}

// IsExamTime 判断是否考试进行中
func IsExamTime(exam *model.ExamDate, now time.Time) bool {
	return now.After(exam.ExamBeginDate) && now.Before(exam.ExamEndDate)
}

// IsExpiredExam 判断考试是否已结束
func IsExpiredExam(exam *model.ExamDate, now time.Time) bool {
	return now.After(exam.ExamEndDate)
}

// GetCountDownString 生成倒计时字符串
func GetCountDownString(exam *model.ExamDate, template string, now time.Time) string {
	if IsExamTime(exam, now) {
		return fmt.Sprintf("%s正在进行中！", exam.ExamDesc)
	}

	if IsExpiredExam(exam, now) {
		return fmt.Sprintf("%s已经结束了。", exam.ExamDesc)
	}

	duration := exam.ExamBeginDate.Sub(now)
	countdownTime := FormatDuration(duration)

	result := template
	result = strings.ReplaceAll(result, "{exam_year}", fmt.Sprintf("%d", exam.ExamYear))
	result = strings.ReplaceAll(result, "{exam}", exam.ExamDesc)
	result = strings.ReplaceAll(result, "{exam_s}", exam.ShortDesc)
	result = strings.ReplaceAll(result, "{time}", countdownTime)

	return result
}

// GetCountDownTime 获取倒计时时间文字
func GetCountDownTime(exam *model.ExamDate, now time.Time) string {
	if IsExamTime(exam, now) || IsExpiredExam(exam, now) {
		return ""
	}

	duration := exam.ExamBeginDate.Sub(now)
	return FormatDuration(duration)
}